<a href="https://colab.research.google.com/github/Tanyajain2006/Fraud-Detection/blob/main/CreditCard_FraudDetection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [37]:
# importing necessary libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

In [3]:
# Loading dataset
dataset = pd.read_csv('/creditcard.csv')

In [4]:
# First 5 rows of dataset
dataset.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [9]:
# Last 5 rows
dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
53566  46047 -2.885416 -2.647887  1.903441  3.726885  3.800798 -1.644480   
53567  46048  1.061410  0.045830  0.970401  1.102883 -0.490685  0.219968   
53568  46048 -0.202740  0.295074  1.924698  1.009344 -0.235931 -0.272232   
53569  46048 -0.730116  0.796820  2.122484 -1.318082  0.019762 -0.857857   
53570  46050  0.791672 -0.522991  0.036747  1.204803 -0.262390 -0.056153   

             V7        V8        V9  ...       V21       V22       V23  \
53566 -1.924198  0.707696 -1.070468  ...  0.284893 -0.268122  0.977417   
53567 -0.424285  0.235960  0.016498  ...  0.033158  0.067127  0.053801   
53568  0.261015 -0.162146 -0.583849  ...  0.321731  1.056540  0.093902   
53569  0.975752 -0.399896  0.359189  ... -0.149736  0.093479 -0.056586   
53570  0.263614 -0.130209  0.125423  ...  0.013131 -0.000000       NaN   

            V24       V25       V26       V27       V28  Amount  Class  
53566  0.136853 -1.144367 -0.535616  0.098494  0.305740   19.02    0.0  
53567 -0.010221  0.209877 -0.484136  0.057583  0.028018   28.70    0.0  
53568  0.608566 -0.931546  0.466156  0.016695 -0.024568   54.81    0.0  
53569  0.650087 -0.473889  0.662115  0.060486 -0.208536    4.53    0.0  
53570       NaN       NaN       NaN       NaN       NaN     NaN    NaN  

[5 rows x 31 columns]

In [10]:
# Dataset information
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53571 entries, 0 to 53570
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    53571 non-null  int64  
 1   V1      53571 non-null  float64
 2   V2      53571 non-null  float64
 3   V3      53571 non-null  float64
 4   V4      53571 non-null  float64
 5   V5      53571 non-null  float64
 6   V6      53571 non-null  float64
 7   V7      53571 non-null  float64
 8   V8      53571 non-null  float64
 9   V9      53571 non-null  float64
 10  V10     53571 non-null  float64
 11  V11     53571 non-null  float64
 12  V12     53571 non-null  float64
 13  V13     53571 non-null  float64
 14  V14     53571 non-null  float64
 15  V15     53571 non-null  float64
 16  V16     53571 non-null  float64
 17  V17     53571 non-null  float64
 18  V18     53571 non-null  float64
 19  V19     53571 non-null  float64
 20  V20     53571 non-null  float64
 21  V21     53571 non-null  float64
 22

In [11]:
# Checking missing values
dataset.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [12]:
# distributions of legit transactions and fraud transactions
dataset['Class'].value_counts()

Class
0.0    53417
1.0      153
Name: count, dtype: int64

This dataset is highly unbalanced

0 represents "Normal Transaction"

1 represents "Fraud Trannsaction"

In [13]:
# seperating the data for analysis
legit = dataset[dataset.Class == 0]
fraud = dataset[dataset.Class == 1]

In [16]:
print(legit.shape)
print(fraud.shape)

(53417, 31)
(153, 31)


In [18]:
# statistical measures of data
print("Legit Amount description \n", legit.Amount.describe())
print("Fraud Amount description \n", fraud.Amount.describe())

Legit Amount description 
 count    53417.000000
mean        94.251260
std        252.353693
min          0.000000
25%          7.680000
50%         25.520000
75%         86.650000
max      12910.930000
Name: Amount, dtype: float64
Fraud Amount description 
 count     153.000000
mean       97.616013
std       230.026856
min         0.000000
25%         1.000000
50%         7.610000
75%        99.990000
max      1809.680000
Name: Amount, dtype: float64


Insights: The mean transaction in fraud is 97.6 while it is 94.25 in case of legit transactions.

So, for all the fraudulent activities, mean transaction is higher than legit ones which is very important factor.

In [20]:
# compare the value for both transactions
dataset.groupby('Class').mean()

Time        V1        V2         V3        V4        V5  \
Class                                                                   
0.0    30011.81811 -0.222505 -0.014659   0.722511  0.163565 -0.242775   
1.0    27472.24183 -7.468948  5.326389 -10.100308  5.793812 -5.528864   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.109604 -0.097131  0.042665  0.104805  ...  0.046251 -0.031568   
1.0   -2.230554 -7.865700  3.617647 -3.492917  ...  0.455720  0.891045   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.106396 -0.038408  0.008094  0.134794  0.019825  0.002289  0.003902   
1.0   -0.218399 -0.277419 -0.084305  0.240593  0.136258  0.586389  0.028185   

          Amount  
Class             
0.0    94.251260  
1.0    97.616013  

[2 rows x 30 columns]

We can see a wide difference between means of legit and fraudulent transactions.

This difference is very important for us and this is how our machine learning algorithm can predict whether the transaction is either legit or fraud one.

Now, dealing with unbalanced data with the process of **under-sampling**.

Build a similar dataset containing similar number of legit and fraudulent transactions. Here we have:

*   Legit = 53417
*   Fraud = 153

So, taking legit sample also of 153 size.

In [22]:
legit_sample = legit.sample(n=153)

In [23]:
# Concatenate two dataframes
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [24]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
17541  28804  1.317783 -0.670148  1.606155 -0.504463 -1.942486 -0.813681   
25357  33572  1.069113 -0.081543  0.645303  1.500371 -0.691602 -0.525919   
44991  42160  1.010190 -0.754970  1.008604  0.345953 -1.392268 -0.380346   
909      685  0.874198 -1.153053  0.591410  0.571892 -1.244914  0.182416   
36649  38629 -0.730634 -0.923464  0.879622  0.598111 -1.907332  0.759404   

             V7        V8        V9  ...       V21       V22       V23  \
17541 -1.241465 -0.056929 -0.370626  ...  0.481826  1.407512 -0.027590   
25357 -0.047113 -0.019396  0.608959  ... -0.247075 -0.738138  0.003144   
44991 -0.579116  0.080283  1.339526  ... -0.181979 -0.545066  0.000247   
909   -0.582385  0.086637 -0.679556  ... -0.088785 -0.127773 -0.287287   
36649  1.600699 -0.405458  0.277858  ...  0.127974  0.751896  0.944554   

            V24       V25       V26       V27       V28  Amount  Class  
17541  0.990530  0.279836 -0.078760  0.065834  0.040607    9.99    0.0  
25357  0.326399  0.426434 -0.531054  0.019032  0.036309   69.25    0.0  
44991  0.438556  0.065121  0.934869 -0.059527  0.030281  112.18    0.0  
909   -0.004026  0.457802 -0.241181  0.033369  0.056633  217.00    0.0  
36649  0.151951 -1.732589  0.981415  0.146812 -0.045594  472.00    0.0  

[5 rows x 31 columns]

In [25]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
50211  44393 -4.617461  3.663395 -5.297446  3.880960 -3.263551 -0.918547   
50537  44532 -0.234922  0.355413  1.972183 -1.255593 -0.681387 -0.665732   
52466  45463 -1.476893  2.122314 -1.229470  1.201849 -0.343264 -1.317704   
52521  45501  1.001992  0.047938 -0.349002  1.493958  0.186939  0.190966   
52584  45541 -1.519244  2.308492 -1.503599  2.064101 -1.000845 -1.016897   

             V7        V8        V9  ...       V21       V22       V23  \
50211 -5.715262  0.831040 -2.457034  ...  2.698175 -0.027081  0.366775   
50537  0.059110 -0.003153  1.122451  ...  0.220670  0.912107 -0.286338   
52466 -1.528142 -0.620953 -1.213040  ...  1.186036 -0.040215 -0.238930   
52521 -0.001112  0.147140  0.580415  ... -0.334417 -1.014315 -0.128427   
52584 -2.059731 -0.275166 -1.562206  ...  1.307871  0.102826 -0.017746   

            V24       V25       V26       V27       V28  Amount  Class  
50211 -0.123011 -0.300457 -0.239996 -0.183463 -0.073360    1.00    1.0  
50537  0.451208  0.188315 -0.531846  0.123185  0.039581    1.00    1.0  
52466  0.110144  0.045418 -0.569232  0.481019 -0.047555    1.00    1.0  
52521 -0.946242  0.456090 -0.453206  0.046627  0.064698  105.99    1.0  
52584  0.149696 -0.096602 -0.369115 -0.019244 -0.208319    1.00    1.0  

[5 rows x 31 columns]

In [26]:
new_dataset['Class'].value_counts()

Class
0.0    153
1.0    153
Name: count, dtype: int64

Insights: Now, we have equal number of both fraud and true transactions.

In [27]:
new_dataset.groupby('Class').mean()

Time        V1        V2         V3        V4        V5  \
Class                                                                    
0.0    30038.666667 -0.219258 -0.326355   0.728595  0.189456 -0.318473   
1.0    27472.241830 -7.468948  5.326389 -10.100308  5.793812 -5.528864   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.021616 -0.091181  0.097077 -0.027749  ...  0.026325 -0.039508   
1.0   -2.230554 -7.865700  3.617647 -3.492917  ...  0.455720  0.891045   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.101406  0.001000 -0.000695  0.154739  0.035697  0.010702 -0.005579   
1.0   -0.218399 -0.277419 -0.084305  0.240593  0.136258  0.586389  0.028185   

           Amount  
Class              
0.0    120.589542  
1.0     97.616013  

[2 rows x 30 columns]

Since, we have equal number of fraud and legit transactions, so we can split the data into features and targets.

In [29]:
X = new_dataset.drop(columns='Class', axis=1)
Y = new_dataset['Class']

print("X \n", X)
print("Y \n", Y)

X 
         Time        V1        V2        V3        V4        V5        V6  \
17541  28804  1.317783 -0.670148  1.606155 -0.504463 -1.942486 -0.813681   
25357  33572  1.069113 -0.081543  0.645303  1.500371 -0.691602 -0.525919   
44991  42160  1.010190 -0.754970  1.008604  0.345953 -1.392268 -0.380346   
909      685  0.874198 -1.153053  0.591410  0.571892 -1.244914  0.182416   
36649  38629 -0.730634 -0.923464  0.879622  0.598111 -1.907332  0.759404   
...      ...       ...       ...       ...       ...       ...       ...   
50211  44393 -4.617461  3.663395 -5.297446  3.880960 -3.263551 -0.918547   
50537  44532 -0.234922  0.355413  1.972183 -1.255593 -0.681387 -0.665732   
52466  45463 -1.476893  2.122314 -1.229470  1.201849 -0.343264 -1.317704   
52521  45501  1.001992  0.047938 -0.349002  1.493958  0.186939  0.190966   
52584  45541 -1.519244  2.308492 -1.503599  2.064101 -1.000845 -1.016897   

             V7        V8        V9  ...       V20       V21       V22  \
17541 -1.

In [33]:
# Split the data into training and testing data.
X_train, X_test, Y_train, Y_test = train_test_split(
    X, Y,
    test_size = 0.2,
    stratify = Y,
    random_state = 2
)

# test_size = 0.2 represents 80% of the data is used for trainig and 20% is used for testing data.

# stratify = Y means stratifying the data based on Y.
# What it means is --> if I don't mention this term, then distribution of 0 and 1 can be very different in training and testing data.
# But when I mention this stratify = Y, then there will be evenly distribution of two classes in both X_training and X_test.

# random_state = 2 means how I want to split the data.

In [34]:
print(X.shape, X_train.shape, X_test.shape)

(306, 30) (244, 30) (62, 30)


Insights: There are about 306 values in original dataset. Out of this, 244 went in training data and 62 went in testing data.

So, we have successfully splitted the data into training and testing data.

In [35]:
# Model Training using Logistic Regression
model = LogisticRegression()

In [36]:
# Training the Logistic Regression model with training data
model.fit(X_train, Y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [40]:
# Model evaluation

# Accuracy on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)
print('Accuracy on Training data :', training_data_accuracy)

# Accuracy on testing data
X_test_prediction = model.predict(X_test)
testing_data_accuracy = accuracy_score(X_test_prediction, Y_test)
print('Accuracy on Testing data :', testing_data_accuracy)

Accuracy on Training data : 0.9754098360655737
Accuracy on Testing data : 0.967741935483871
